File: Context/GameSpec.Base/_LIB/Compression/Lzss.cs
using System;
using System.IO;

namespace Compression
{
    internal class Lzss
    {
        internal class BinaryReaderE : BinaryReader
        {
            public BinaryReaderE(Stream stream) : base(stream) { }
            public override short ReadInt16() { var numArray = ReadBytes(2); Array.Reverse((Array)numArray); return BitConverter.ToInt16(numArray, 0); }
            public override int ReadInt32() { var numArray = ReadBytes(4); Array.Reverse((Array)numArray); return BitConverter.ToInt32(numArray, 0); }
            public override uint ReadUInt32() { var numArray = ReadBytes(4); Array.Reverse((Array)numArray); return BitConverter.ToUInt32(numArray, 0); }
            public uint ReadUInt() => base.ReadUInt32();
        }

        int uncompressedSize;
        BinaryReaderE stream;
        const short DICT_SIZE = 4096;
        const short MIN_MATCH = 3;
        const short MAX_MATCH = 18;
        byte[] output;
        byte[] dictionary;
        short NR;
        short DO;
        short DI;
        int OI;

        public Lzss(BinaryReaderE stream, int uncompressedSize)
        {
            this.stream = stream;
            this.uncompressedSize = uncompressedSize;
        }

        bool LastByte() => stream.BaseStream.Position == stream.BaseStream.Length;

        void ClearDict()
        {
            for (var index = 0; index < DICT_SIZE; ++index) dictionary[index] = 32;
            DI = DICT_SIZE - MAX_MATCH;
        }

        byte ReadByte() { ++NR; return stream.ReadByte(); }

        byte[] ReadBytes(int bytes) => stream.ReadBytes(bytes);

        void WriteByte(byte b)
        {
            output[OI++] = b;
            dictionary[DI++ % DICT_SIZE] = b;
        }

        void WriteBytes(byte[] bytes)
        {
            foreach (var num in bytes)
            {
                if (OI >= uncompressedSize) break;
                output[OI++] = num;
            }
        }

        byte ReadDict() => dictionary[DO++ % DICT_SIZE];

        short ReadInt16() => stream.ReadInt16();

        void ReadBlock(int N)
        {
            NR = 0;
            if (N < 0) WriteBytes(ReadBytes(N * -1));
            else
            {
                ClearDict();
                while (NR < N && !LastByte())
                {
                    var num1 = ReadByte();
                    if (NR >= N || LastByte()) break;
                    for (var index1 = 0; index1 < 8; ++index1)
                    {
                        if (num1 % 2 == 1)
                        {
                            WriteByte(ReadByte());
                            if (NR >= N) return;
                        }
                        else
                        {
                            if (NR >= N) return;
                            DO = ReadByte();
                            if (NR >= N) return;
                            var num2 = ReadByte();
                            DO |= (short)((num2 & 240) << 4);
                            var num3 = ((int)num2 & 15) + MIN_MATCH;
                            for (var index2 = 0; index2 < num3; ++index2) WriteByte(ReadDict());
                        }
                        num1 >>= 1;
                        if (LastByte()) return;
                    }
                }
            }
        }

        public byte[] Decompress()
        {
            output = new byte[uncompressedSize];
            dictionary = new byte[DICT_SIZE];
            while (!LastByte())
            {
                var N = ReadInt16();
                if (N != 0) ReadBlock(N);
                else break;
            }
            return output;
        }
    }
}

File: Python/gamex/_LIB/compression/lzss.py
import os
from io import BytesIO

DICT_SIZE = 4096
MIN_MATCH = 3
MAX_MATCH = 18

class Lzss:
    def __init__(self, stream: BytesIO, uncompressedSize: int):
        self.stream = stream
        self.uncompressedSize = uncompressedSize
        self.streamLength = stream.seek(0, os.SEEK_END)
        stream.seek(0, os.SEEK_SET)
        self.out = bytearray()
        self.dict = bytearray()
        self.NR = 0
        self.DO = 0
        self.DI = 0
        self.OI = 0
    
    def _readByte(self) -> int: self.NR += 1; return int.from_bytes(self.stream.read(1), 'big', signed=False)
    def _readInt16(self) -> int: return int.from_bytes(self.stream.read(2), 'big', signed=True)
    def _readDict(self) -> int: ret = self.dict[self.DO % DICT_SIZE]; self.DO+=1; return ret
    def _lastByte(self) -> bool: return self.stream.tell() == self.streamLength
    def _readBlock(self, N):
        s = self.stream
        self.NR = 0
        if N < 0: self._writeBytes(s.read(N * -1))
        else:
            self._clearDict()
            while self.NR < N and not self._lastByte():
                num1 = self._readByte()
                if self.NR >= N or self._lastByte(): break
                for index1 in range(8):
                    if (num1 % 2) == 1:
                        self._writeByte(self._readByte())
                        if self.NR >= N: return
                    else:
                        if self.NR >= N: return
                        self.DO = self._readByte()
                        if self.NR >= N: return
                        num2 = self._readByte()
                        self.DO |= ((num2 & 240) << 4) & 0xFFFF
                        num3 = (num2 & 15) + MIN_MATCH
                        for index2 in range(num3): self._writeByte(self._readDict())
                    num1 >>= 1
                    if self._lastByte(): return
    def _clearDict(self) -> None:
         for index in range(DICT_SIZE): self.dict[index] = 32
         self.DI = DICT_SIZE - MAX_MATCH
    def _writeByte(self, b) -> None: 
        self.out[self.OI] = b; self.OI += 1
        self.dict[self.DI % DICT_SIZE] = b; self.DI += 1
    def _writeBytes(self, b) -> None:
        for num in b:
            if self.OI >= self.uncompressedSize: break
            self.out[self.OI] = num; self.OI += 1
    def decompress(self) -> bytearray:
        self.out = bytearray(self.uncompressedSize)
        self.dict = bytearray(DICT_SIZE)
        while not self._lastByte():
            N = self._readInt16()
            if N != 0: self._readBlock(N)
            else: break
        return self.out